### ISSUE TITLE
biquadratic coupling

### ISSUE THREAD

**Issue by SediE7** [ubermag/help#261]
Hi dear all
@fangohr  @JLeliaert  @marijanbeg  @swapneelap @lang-m @
I try to simulate STNO with both bilinear and biquadratic coupling constant (this paper: **Ultra‑wide‑band millimeter‑wave
generator using spin torque oscillator with strong interlayer exchange couplings    https://doi.org/10.1038/s41598-022-15014-y**)
But neither OOMMF nor mumax3 have any native support for simulations involving the biquadratic coupling constant, how can I simulate this STNO with both bilinear and biquadratic coupling constant?

Best regards,
Sedi Emad

**Comment by SediE7** [comment]
Hi @samjrholt
Thank you for your response.
Best regards,
Sedi Emad

**Comment by jedcheng** [comment]
Actually, I am from the same research group in the paper (but I only joined 3 weeks ago)
Are there any updates from your attempts? @SediE7 

I am also trying to reproduce the results from oommf.
After reviewing the custom LLG equation solver used in the group, I believe the RKKY term in ubermag provides the API to do the coupling.

However, I encountered a problem. (I have been using mumax for years but very new to oommf)
Is it possible to apply a non-polarized current so that I can pass it through my custom fixed layer?

**Comment by swapneelap** [comment]
@jedcheng Thank you for pointing out the RKKY energy for the description of bilinear and biquadratic exchange coupling between two layers separated by a spacer layer. If I understand correctly, you want to study spin-transfer torque in multi-layers using the Slonczewski model? If that is the case, it is assumed that passing a non-polarised current through a pinned layer essentially picks-up the magnetisation of the pinned layer before reaching the free layer. Hence, to simulate the free layer, you need to define the `mp` parameter (orientation of the spin polarisation of the current), which is same as the orientation of magnetisation of the pinned layer, in the `Slonczewski` dynamics term in Ubermag. The documentation of the same for OOMMF can be found here: https://math.nist.gov/oommf/doc/userguide20a3/userguide/Standard_Oxs_Ext_Child_Clas.html#SX

**Comment by jedcheng** [comment]
@swapneelap Thank you for the reply. 

My idea was to simulate both the fixed layer and the free layer in OOMMF. 
After a weekend of trial and error, I settled into simulating both a Slonczewski torque from a polarized current and a very thin fixed layer. The very thin fixed layer's sole purpose is to provide the biquadratic coupling needed.

I was able to increase the current density by 10 fold which the STO could still operate. The oscillating frequency increased 4x to 140GHz compared to the simualtions without the RKKY term. 

My purpose was to verify that the same effects could be observed outside of the solver created by our adjacent research group. Thus the results are good enough for me.
